usr/src/uts/common/io/scsi/adapters/smartpqi/smartpqi_util.c

```diff
@@ -0,0 +1,947 @@
+static char *cmd_state_str(pqi_cmd_state_t state);
+static void dump_raid(pqi_state_t s, void *v, pqi_index_t idx);
+static void dump_aio(void *v);
+static void show_error_detail(pqi_state_t s);
+static void cmd_finish_task(void *v);
+
+/*
+ * []------------------------------------------------------------------[]
+ * | Entry points for this file						|
+ * []------------------------------------------------------------------[]
+ */
+
+int
+pqi_is_offline(pqi_state_t s)
+{
+	return (s->s_offline);
+}
+
+/*
+ * pqi_alloc_io -- return next available slot.
+ */
+pqi_io_request_t *
+pqi_alloc_io(pqi_state_t s)
+{
+	pqi_io_request_t	*io	= NULL;
+	uint16_t		loop;
+	uint16_t		i;
+
+	mutex_enter(&s->s_io_mutex);
+	i = s->s_next_io_slot; /* just a hint */
+	s->s_io_need++;
+	for (;;) {
+		for (loop = 0; loop < s->s_max_io_slots; loop++) {
+			io = &s->s_io_rqst_pool[i];
+			i = (i + 1) % s->s_max_io_slots;
+			if (io->io_refcount == 0) {
+				io->io_refcount = 1;
+				break;
+			}
+		}
+		if (loop != s->s_max_io_slots)
+			break;
+
+		s->s_io_had2wait++;
+		s->s_io_wait_cnt++;
+		if (cv_wait_sig(&s->s_io_condvar, &s->s_io_mutex) == 0) {
+			s->s_io_sig++;
+			io = NULL;
+			break;
+		}
+		i = s->s_next_io_slot; /* just a hint */
+	}
+	s->s_next_io_slot = i;
+	mutex_exit(&s->s_io_mutex);
+
+	if (io != NULL)
+		reinit_io(io);
+	return (io);
+}
+
+void
+pqi_free_io(pqi_io_request_t *io)
+{
+	pqi_state_t	s = io->io_softc;
+
+	mutex_enter(&s->s_io_mutex);
+	ASSERT(io->io_refcount == 1);
+	io->io_refcount = 0;
+	reinit_io(io);
+	if (s->s_io_wait_cnt != 0) {
+		s->s_io_wait_cnt--;
+		cv_signal(&s->s_io_condvar);
+	}
+	mutex_exit(&s->s_io_mutex);
+}
+
+
+void
+pqi_dump_io(pqi_io_request_t *io)
+{
+	pqi_iu_header_t	*hdr = io->io_iu;
+	pqi_state_t	s;
+
+	if (io->io_cmd != NULL) {
+		s = io->io_cmd->pc_softc;
+	} else {
+		/*
+		 * Early on, during driver attach, commands are run without
+		 * a pqi_cmd_t structure associated. These io requests are
+		 * low level operations direct to the HBA. So, grab a
+		 * reference to the first and only instance through the
+		 * DDI interface. Even though there might be multiple HBA's
+		 * grabbing the first is okay since dump_raid() only references
+		 * the debug level which will be the same for all the
+		 * controllers.
+		 */
+		s = ddi_get_soft_state(pqi_state, 0);
+	}
+
+	if (hdr->iu_type == PQI_REQUEST_IU_AIO_PATH_IO) {
+		dump_aio(io->io_iu);
+	} else if (hdr->iu_type == PQI_REQUEST_IU_RAID_PATH_IO) {
+		dump_raid(s, io->io_iu, io->io_pi);
+	}
+}
+
+/*
+ * pqi_cmd_sm -- state machine for command
+ *
+ * NOTE: PQI_CMD_CMPLT and PQI_CMD_FATAL will drop the pd_mutex and regain
+ * it even if grab_lock==B_FALSE.
+ */
+void
+pqi_cmd_sm(pqi_cmd_t cmd, pqi_cmd_state_t new_state, boolean_t grab_lock)
+{
+	pqi_device_t	devp = cmd->pc_device;
+	pqi_state_t	s = cmd->pc_softc;
+
+	if (cmd->pc_softc->s_debug_level & DBG_LVL_STATE) {
+		cmn_err(CE_NOTE, "%s: cmd=%p (%s) -> (%s)\n", __func__,
```

Comment:
Probably in other places too -- cmn_err(CE_NOTE, ...) should already appends a newline, you don't need to add one yourself.  (The extra newline actually can make it harder to debug in some circumstances, because more messages get scrolled off the screen when using a VGA console.  This is a pet peeve of mine.)